File: 1_stack/单调栈/rangeMinChange.ts
interface IChange {
  /** 区间左端点开始位置 */
  lStart: number
  /** 区间左端点结束位置 */
  lEnd: number
  /** 旧的最小值 */
  oldMin: number
  /** 新的最小值 */
  newMin: number
}

/**
 * 维护区间最小值的变化历史。
 * 返回：res[i]，表示右端点r=i+1时，所有受影响区间[l,r)的最小值变化记录：(l, r, old_min, new_min)
 * 每次右端点推进，所有被当前元素“刷新”最小值的区间都会被记录下来，适用于区间DP、单调栈优化等场景。
 */
function rangeMinChange(nums: ArrayLike<number>, defaultValue = Infinity): IChange[][] {
  const n = nums.length
  const res: IChange[][] = Array(n)
  for (let i = 0; i < n; i++) res[i] = []

  type Triple = {
    l: number
    r: number
    min: number
  }

  const stack: Triple[] = []
  for (let i = 0; i < n; i++) {
    const v = nums[i]
    res[i].push({ lStart: i, lEnd: i + 1, oldMin: defaultValue, newMin: v })
    let ptr = i
    while (stack.length > 0) {
      const lc = stack[stack.length - 1]
      if (lc.min <= v) break
      res[i].push({ lStart: lc.l, lEnd: lc.r, oldMin: lc.min, newMin: v })
      ptr = lc.l
      stack.pop()
    }
    stack.push({ l: ptr, r: i + 1, min: v })
    res[i].reverse()
  }

  return res
}

export { rangeMinChange }

if (typeof require !== 'undefined' && typeof module !== 'undefined' && require.main === module) {
  // Test cases
  const nums = [3, 1, 4, 1, 5, 9, 2, 6]
  const changes = rangeMinChange(nums)

  for (let i = 0; i < changes.length; i++) {
    console.log(`Right endpoint r=${i + 1}:`, changes[i])
  }
}

File: 1_stack/单调栈/range_min_change.py
from typing import List, Tuple

INF = int(1e18)


def range_min_change(nums: List[int], default_val=INF) -> List[List[Tuple[int, int, int, int]]]:
    """
    维护区间最小值的变化历史。
    返回：res[i]，表示右端点r=i+1时，所有受影响区间[l,r)的最小值变化记录：(l, r, old_min, new_min)
    每次右端点推进，所有被当前元素“刷新”最小值的区间都会被记录下来，适用于区间DP、单调栈优化等场景。
    """
    n = len(nums)
    res = [[] for _ in range(n)]
    stack = []
    for i, v in enumerate(nums):
        res[i].append((i, i + 1, default_val, v))
        ptr = i
        while stack:
            l, r, x = stack[-1]
            if x <= v:
                break
            res[i].append((l, r, x, v))
            ptr = l
            stack.pop()
        stack.append((ptr, i + 1, v))
        res[i].reverse()
    return res


if __name__ == "__main__":
    # Example usage
    # Changes for index 0: [(0, 1, 1000000000000000000, 5)]
    # Changes for index 1: [(0, 1, 5, 3), (1, 2, 1000000000000000000, 3)]
    # Changes for index 2: [(2, 3, 1000000000000000000, 4)]
    # Changes for index 3: [(0, 2, 3, 2), (2, 3, 4, 2), (3, 4, 1000000000000000000, 2)]
    # Changes for index 4: [(4, 5, 1000000000000000000, 6)]
    A = [5, 3, 4, 2, 6]
    result = range_min_change(A)
    for i, changes in enumerate(result):
        print(f"Changes for index {i}: {changes}")
